compiler/testData/codegen/box/callableReference/serializability/adaptedReferences.kt

```diff
@@ -1,3 +1,4 @@
+// IGNORE_BACKEND_FIR: JVM_IR
```

Comment:
Do you know why this test became broken?

Reply:
When adding an adaptee call inside an adapter function, transient receiver (`A` in `A::foo`) is not handled well, and I left a TODO.

For example, the last one `allOfTheAbove(A::foo)`, the expected IR dump looks like:
```
CALL 'public final fun check (x: kotlin.Any): kotlin.Unit declared in <root>.AdaptedReferencesKt' type=kotlin.Unit origin=null
          x: CALL 'public final fun allOfTheAbove (f: kotlin.Function1<<root>.A, kotlin.Unit>): kotlin.Any declared in <root>.AdaptedReferencesKt' type=kotlin.Any origin=null
            f: BLOCK type=kotlin.Function1<<root>.A, kotlin.Unit> origin=ADAPTED_FUNCTION_REFERENCE
              FUN ADAPTER_FOR_CALLABLE_REFERENCE name:foo visibility:local modality:FINAL <> (p0:<root>.A) returnType:kotlin.Unit
                VALUE_PARAMETER ADAPTER_PARAMETER_FOR_CALLABLE_REFERENCE name:p0 index:0 type:<root>.A
                BLOCK_BODY
                  TYPE_OP type=kotlin.Unit origin=IMPLICIT_COERCION_TO_UNIT typeOperand=kotlin.Unit
                    CALL 'public final fun foo (s: kotlin.String, vararg xs: kotlin.Long): kotlin.String declared in <root>.A' type=kotlin.String origin=null
                      $this: GET_VAR 'p0: <root>.A declared in <root>.AdaptedReferencesKt.box.foo' type=<root>.A origin=null
              FUNCTION_REFERENCE 'local final fun foo (p0: <root>.A): kotlin.Unit declared in <root>.AdaptedReferencesKt.box' type=kotlin.Function1<<root>.A, kotlin.Unit> origin=ADAPTED_FUNCTION_REFERENCE reflectionTarget=null
```
v.s. `fir` result with this PR looks like:
```
CALL 'public final fun check (x: kotlin.Any): kotlin.Unit declared in <root>.AdaptedReferencesKt' type=kotlin.Unit origin=null
          x: CALL 'public final fun allOfTheAbove (f: kotlin.Function1<<root>.A, kotlin.Unit>): kotlin.Any declared in <root>.AdaptedReferencesKt' type=kotlin.Any origin=null
            f: BLOCK type=kotlin.Function1<<root>.A, kotlin.Unit> origin=ADAPTED_FUNCTION_REFERENCE
              FUN ADAPTER_FOR_CALLABLE_REFERENCE name:foo visibility:local modality:FINAL <> (p0:<root>.A) returnType:kotlin.Unit
                VALUE_PARAMETER ADAPTER_PARAMETER_FOR_CALLABLE_REFERENCE name:p0 index:0 type:<root>.A
                BLOCK_BODY
                  CALL 'public final fun foo (s: kotlin.String, vararg xs: kotlin.Long): kotlin.String declared in <root>.A' type=kotlin.String origin=null
              FUNCTION_REFERENCE 'local final fun foo (p0: <root>.A): kotlin.Unit declared in <root>.AdaptedReferencesKt.box' type=kotlin.Function1<<root>.A, kotlin.Unit> origin=ADAPTED_FUNCTION_REFERENCE reflectionTarget=null
```
The missing part is the bound receiver. I found that a class symbol `A` is set as an explicit receiver, but not converted to IR receiver. Then, we give up setting the receiver for the adaptee call.

I know, a real question is, why it worked when we _don't_ apply this adapted reference conversion to coercion-to-unit case. That's still mysterious, and I don't fully understand how reference lowering in the backend deals with adapted reference... On the other hand, another real question is, when we really apply reference conversion to coercion-to-unit? I don't know either... :( `ReflectionReferenceGenerator` in `psi2ir` doesn't have good answers either.

Reply:
Will be resolved at https://github.com/JetBrains/kotlin/pull/3630